src/collision/cm_trace.c

```diff
@@ -414,130 +371,112 @@ next:;
+		data->trace.end = data->end;
 	} else {
-		data.trace.end = Vec3_Mix(data.start, data.end, data.trace.fraction);
+		data->trace.end = Vec3_Mix(data->start, data->end, data->trace.fraction);
 	}
-
-	return data.trace;
 }
 
 /**
  * @brief Primary collision detection entry point. This function recurses down
  * the BSP tree from the specified head node, clipping the desired movement to
- * brushes that match the specified contents mask.
+ * brushes that match the specified contents mask. The matrix and inverse matrix
+ * are used to transform the trace input and output, allowing for the trace to
+ * test against a translated or rotated BSP tree.
  *
  * @param start The starting point.
  * @param end The desired end point.
  * @param bounds The bounding box, in model space.
- * @param head_node The BSP head node to recurse down. For inline BSP models,
- * the head node is the root of the model's subtree. For mesh models, a
- * special reserved box hull and head node are used.
+ * @param head_node The head node of the inline model or box hull.
  * @param contents The contents mask to clip to.
- * @param matrix The matrix to adjust tested planes by.
- * @param inverse_matrix The inverse matrix to adjust the inputs by.
+ * @param matrix The matrix of the inline model or box hull.
+ * @param inverse_matrix The inverse matrix of the inline model or box hull.
  *
  * @return The trace.
  */
-cm_trace_t Cm_TransformedBoxTrace(const vec3_t start, const vec3_t end, const box3_t bounds, int32_t head_node,
-					              int32_t contents, const mat4_t matrix, const mat4_t inverse_matrix) {
-
-	data.start = start;
-	data.end = end;
-	data.bounds = bounds;
-	data.head_node = head_node;
-	data.matrix = matrix;
-	data.inverse_matrix = inverse_matrix;
-	data.abs_bounds = Cm_TraceBounds(start, end, bounds);
-	data.contents = contents;
-	data.is_transformed = true;
-	data.trace = (cm_trace_t) {
-		.fraction = 1.f
+cm_trace_t Cm_TransformedBoxTrace(const vec3_t start, const vec3_t end, const box3_t bounds,
+								  int32_t head_node, int32_t contents,
+								  const mat4_t matrix, const mat4_t inverse_matrix) {
+
+	cm_trace_data_t data = {
+		.start = Mat4_Transform(inverse_matrix, start),
+		.end = Mat4_Transform(inverse_matrix, end),
+		.bounds = bounds,
+		.head_node = head_node,
+		.contents = contents
 	};
+
+	data.size = Box3_Symetrical(Box3_Expand(bounds, BOX_EPSILON));
+	data.abs_bounds = Cm_TraceBounds(data.start, data.end, data.bounds);
+
+	Box3_ToPoints(data.bounds, data.offsets);
+	memset(data.brush_cache, 0xff, sizeof(data.brush_cache));
+
 	data.unnudged_fraction = 1.f + TRACE_EPSILON;
+	data.trace.fraction = 1.f;
+
+	Cm_BoxTrace_(&data);
+
+	data.trace.end = Mat4_Transform(matrix, data.trace.end);
 
-	return Cm_BoxTrace_();
+	const vec4_t p = Mat4_TransformPlane(matrix, data.trace.plane.normal, data.trace.plane.dist);
+	data.trace.plane.normal = Vec4_XYZ(p);
+	data.trace.plane.dist = p.w;
+	
+	return data.trace;
```

Comment:
Un-transform start and end, and use those to calculate abs_bounds so that it too is un-transformed. Then run the trace. Finally, transform the trace end point, and the impacted plane.